R/parallelization.R

```diff
@@ -13,97 +68,106 @@
-#' environment. This helps to prevent inadvertently shipping large data objects
-#' to daemons, where they are not needed.
+#' \pkg{purrr}'s parallelization is powered by \CRANpkg{mirai}. See the
+#' [mirai website](https://mirai.r-lib.org/) for more details.
 #'
-#' Any non-package function supplied to `.f` will be automatically crated. When
-#' this happens, a confirmation along with the crate size is printed to the
-#' console. Package functions are not crated as these are already
-#' self-contained.
+#' @export
+#' @examplesIf interactive() && requireNamespace("mirai", quietly = TRUE) && requireNamespace("carrier", quietly = TRUE)
+#' # Run in interactive sessions only as spawns additional processes
 #'
-#' If your function `.f()` contains free variables, for example it references
-#' other local functions in its body, then explicitly [carrier::crate()] your
-#' function supplying these variables to its `...` argument. This ensures that
-#' these objects are available to `.f()` when it is executed in a parallel
-#' process.
+#' # To use parallelized map:
+#' # 1. Set daemons (number of parallel processes) first:
+#' mirai::daemons(2)
 #'
-#' Examples:
-#' \preformatted{
-#' # package functions are not auto-crated:
-#' map(1:3, stats::runif, .parallel = TRUE)
+#' # 2. Wrap .f with in_parallel():
+#' mtcars |> map_dbl(in_parallel(\(x) sum(x)))
 #'
-#' # other functions (incl. anonymous functions) are auto-crated:
-#' mtcars |> map_dbl(function(...) sum(...), .parallel = TRUE)
+#' 1:10 |>
+#'   map(in_parallel(\(x) stats::rnorm(10, mean = x))) |>
+#'   map_dbl(in_parallel(\(x) mean(x)))
 #'
-#' # explicitly crate a function to include other objects required by it:
-#' fun <- function(x) \{x + x \%\% 2 \}
-#' map(1:3, carrier::crate(function(x) x + fun(x), fun = fun), .parallel = TRUE)
+#' # A locally-defined function should be passed via ... of in_parallel():
+#' slow_lm <- \(formula, data) {
+#'   Sys.sleep(0.5)
+#'   lm(formula, data)
 #' }
 #'
-#' For details on further options, see [carrier::crate].
+#' mtcars |>
+#'   split(mtcars$cyl) |>
+#'   map(in_parallel(\(df) slow_lm(mpg ~ disp, data = df), slow_lm = slow_lm))
 #'
-#' # Further documentation
-#'
-#' \pkg{purrr}'s parallelization is powered by \CRANpkg{mirai}. See the
-#' [mirai website](https://mirai.r-lib.org/) for more details.
-#'
-#' Crating is provided by the \CRANpkg{carrier} package. See the
-#' [carrier readme](https://github.com/r-lib/carrier) for more details.
+#' # Tear down daemons when no longer in use:
+#' mirai::daemons(0)
 #'
-#' @name parallelization
-NULL
+in_parallel <- function(.f, ...) {
+  if (is.null(the$packages_installed)) {
+    check_installed(
+      c("carrier", "mirai"),
+      version = c("0.1.1", "2.3.0"),
+      reason = "for parallel map."
+    )
+    the$packages_installed <- TRUE
+  }
+  inject(carrier::crate(!!substitute(.f), !!!list(...)))
+}
+
+is_crate <- function(x) {
+  inherits(x, "crate")
```

Comment:
I must have created carrier before we started to systematically namespace S3 classes with package names...